usersession/agent/session_agent.go

```diff
@@ -50,6 +52,26 @@ type SessionAgent struct {
 	IdleTimeout time.Duration
 }
 
+const sessionAgentBusName = "io.snapcraft.SessionAgent"
+
+func dbusSessionBus() (*dbus.Conn, error) {
+	// use a private connection to the session bus, this way we can manage
+	// its lifetime without worrying of breaking other code
+	conn, err := dbus.SessionBusPrivate()
```

Comment:
I wonder what happens on systems that have systemd session services but do not have dbus in session services? My reading of the DBus go code seems to suggest this will spawn a throw-away daemon that is not managed by systemd.

Reply:
Ubuntu 16.04 is probably the only interesting case here: everything else I've seen either lacks a user instance of systemd (e.g. CentOS 7 where it has been disabled), or has both a user instance of systemd and the session bus managed by systemd.

In the specific case of Ubuntu 16.04, the Upstart post-start script writes the session bus address to `$XDG_RUNTIME_DIR/dbus-session`, which godbus checks for.  There is a small chance that we will be started via socket activation before Upstart starts the session bus, but I'm not sure what we can do here.

Reply:
I think some server environments also have one but not the other. IIRC this is only fixed relatively recently and our test suite is masking that by adding the extra packages.

Reply:
The Debian packages definitely split the unit files out into a separate `dbus-user-session` package, so presumably it is possible to get into this situation on other distro releases too.

I don't think we have Upstart running for ssh sessions into 16.04 either, so I agree that it would make sense to treat this as a soft error and log it rather than causing the session agent to fail.  REST API calls that want to post notifications can easily check to see if cmd.s.bus is non nil to see whether the session bus is available.

Reply:
yes, making this not fatal seems a good idea if we can degrade other specific behavior instead

Reply:
@jhenstridge do you think `dbusutil.SessionBus()` should have a twin function for a private variant? There we explicitly do not start dbus-daemon if one is not already running.

Reply:
In addition, the whole `dbusSessionBus` function feels like it should be moved to `dbusutil` package, as it's not specific to the user session agent.

Reply:
I mainly picked a private connection because that's what userd was using.  The main benefits of a private connection are that you are free to close them at a point convenient to you.  Whereas, the shared connection could be in use by any other library.

As for the helper, Ideally we'd just upgrade to a newer version of go-dbus, where it could be replaced by `dbus.ConnectSessionBus()`.  We'd need to check distros that don't use our vendored versions though.